compiler/rustc_mir/src/transform/inline.rs

```diff
@@ -115,7 +124,10 @@ impl Inliner<'tcx> {
                     // in the other direction.
                     if !self.tcx.dep_graph.is_fully_enabled()
                         && self_hir_id < callee_hir_id
-                        && caller_body.generator_kind.is_none()
+                        // Or we know for sure that the function we're calling does not recurse.
+                        || !self.tcx.mir_inliner_cycle(callee_def_id)
+                        // Or we know for sure that we don't recurse.
+                        || !self_recurses
```

Comment:
Could you move `|| !self_recurses` above `|| !self.tcx.mir_inliner_cycle(callee_def_id)`? That prevents an invocation of `mir_inliner_cycle` when `!self_recurses`.